src/main/java/com/aws/greengrass/deployment/activator/KernelUpdateActivator.java

```diff
@@ -120,10 +126,13 @@ void rollback(Deployment deployment, Throwable failureCause) {
         deployment.setErrorTypes(errorReport.getRight());
         deployment.setStageDetails(Utils.generateFailureMessage(failureCause));
 
-        final boolean bootstrapOnRollbackRequired = kernelAlternatives.prepareBootstrapOnRollbackIfNeeded(
-                kernel.getContext(), deploymentDirectoryManager, bootstrapManager);
+        // Persist deployment metadata only if restart is required
+        if (requiresRestart) {
```

Comment:
QQ: when bootstrap prepared failed, and we don't persist the deployment metadata, if we use systemctl to restart Nucleus, would it still look for deployment metadata when restarting?

Reply:
We will still persist deployment metadata but only the launch directory of the deployment that failed. After rollback, the launch directory will revert back to previous one (before current deployment was received). So after restart, yes, Nucleus will still look for deployment metadata but its fine if it does not find it and will launch in default mode.